### ISSUE TITLE
Rule Request: discouraged_decimal_float_literal

### ISSUE THREAD

**Issue by johnhammerlund** [realm/SwiftLint#2833]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

#### Description

This rule would trigger a violation when attempting to initialize a `Decimal` via float literal.

> 1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

Normally, I wouldn't be quick to file a request based on a Swift bug, but this one has gone unfixed for 3 years (and has existed as long as the `ExpressibleByXYZ` & their predecessors have been around). With ABI stability, the likelihood of this being fixed in the foreseeable future has diminished.

**To boil it down:** `Decimal` conforms by `ExpressibleByFloatLiteral`. Idiomatically, the syntax makes perfect sense:
```swift
let myDecimal: Decimal = 0.59
```
Since `Decimal` exists for the sake of accurate base10 arithmetic, we would expect to never encounter rounding errors:
```swift
let formatter = NumberFormatter()
formatter.maximumFractionDigits = 2
formatter.minimumFractionDigits = 2
formatter.roundingMode = .floor
let formatted = formatter.string(from: NSDecimalNumber(decimal: nonPreciseDecimal))

formatted == "0.59" // Wrong, it's 0.58
```
This is because of how `ExpressibleByFloatLiteral` works: the implementor must construct the type via `Double`. This means that any time a `Decimal` is declared as a literal, it is initially represented in binary before it's converted to base10, which introduces rounding errors that could be extremely easy to miss.

Relevant Discussion (includes related bugs): https://forums.swift.org/t/exact-nsdecimalnumber-via-literal/12834

> 2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

Would trigger:
```swift
let decimal: Decimal = 0.59 // violation
let anotherDecimal: Decimal
anotherDecimal = 0.59 // violation
```

Would not trigger:
```swift
let decimal = Decimal(string: "0.59")
let anotherDecimal = Decimal(sign: .plus, exponent: -2, significand: 59)
let wholeDecimal: Decimal = 5
```

> 3. Should the rule be configurable, if so what parameters should be configurable?

Nope

> 4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Only depending on how complex the implementation might be (in which case more false positives might arise), then this should be enabled by default. Any case that declares a `Decimal` via float literal has likely nullified the benefits of using `Decimal` in the first place.

EDIT: To truly catch all major cases, this will likely require type-checked AST. It may have to remain Opt-In while `analyze` remains experimental.
